File: apps/browser-extension/src/mailbox.ts
// Semantic provider data only. This module cannot read DOM, click, navigate,
// mint an account binding, or replace C4's actual Chrome ownership checks.
import * as p from "./protocol.js";

export type Message = {
  id: string; thread: string; timestamp_ms: number; sender: string;
  subject: string; reference: string; body: string;
  body_complete: true; thread_expanded: true; inbox_message_id:string;
};
export type Batch = {
  binding_revision: string; account: string; scope: "inbox";
  document: string; dom_revision: number; ordinal: number;
  cursor: string | null; next: string | null; end_of_inbox: boolean;
  order: "individual_messages_newest_first"; messages: Message[]; unseen_at_most_ms:number|null;
};

const encoder = new TextEncoder();
function identity(value: unknown, maximum: number): value is string {
  return typeof value === "string" && value.length > 0 && value.length <= maximum
    && /^[\x21-\x7e]+$/.test(value);
}
function text(value: unknown, maximum: number, multiline = false): value is string {
  return typeof value === "string" && value.length <= maximum
    && encoder.encode(value).length <= maximum
    && !/[\p{Cc}\p{Cs}]/u.test(multiline ? value.replace(/[\n\t]/g, "") : value);
}
function message(value: unknown): value is Message {
  return p.object(value, ["id", "thread", "timestamp_ms", "sender", "subject", "reference", "body", "body_complete", "thread_expanded", "inbox_message_id"])
    && identity(value.id, 128) && identity(value.thread, 128)
    && p.counter(value.timestamp_ms) && value.timestamp_ms <= 8_640_000_000_000_000
    && text(value.sender, 320) && !!value.sender.trim() && text(value.subject, 512)
    && identity(value.reference, 2048) && text(value.body, 65536, true)
    && value.body_complete === true && value.thread_expanded === true && value.inbox_message_id===value.id;
}

export function batch(value: unknown, requested: number): value is Batch {
  if (!Number.isInteger(requested) || requested < 1 || requested > 100
    || !p.object(value, ["binding_revision", "account", "scope", "document", "dom_revision", "ordinal", "cursor", "next", "end_of_inbox", "order", "messages", "unseen_at_most_ms"])
    || !p.id(value.binding_revision) || !identity(value.account, 320) || value.scope !== "inbox"
    || !identity(value.document, 128) || !p.counter(value.dom_revision)
    || !Number.isInteger(value.ordinal) || (value.ordinal as number) < 0 || (value.ordinal as number) >= 32
    || !(value.cursor === null || identity(value.cursor, 256))
    || !(value.next === null || identity(value.next, 256))
    || typeof value.end_of_inbox !== "boolean" || value.end_of_inbox === (value.next !== null)
    || value.end_of_inbox !== (value.unseen_at_most_ms===null)
    || !(value.unseen_at_most_ms===null || (p.counter(value.unseen_at_most_ms)&&value.unseen_at_most_ms<=8_640_000_000_000_000))
    || value.order !== "individual_messages_newest_first" || !Array.isArray(value.messages)
    || value.messages.length > requested || (!value.end_of_inbox && value.messages.length === 0)) return false;
  const ids = new Set<string>();
  let previous = Number.MAX_SAFE_INTEGER, bytes = 0;
  for (const item of value.messages) {
    if (!message(item) || ids.has(item.id) || item.timestamp_ms > previous) return false;
    previous = item.timestamp_ms; ids.add(item.id); bytes += encoder.encode(item.body).length;
    if (bytes > 1048576) return false;
  }
  return value.unseen_at_most_ms===null || (value.messages.length>0 && Number(value.unseen_at_most_ms)<previous);
}

// Readiness is a separate provider observation. A URL, profile selection or
// installed extension cannot set authenticated/composeEnabled on its own.
export type Readiness = {
  binding_revision: string; document: string; dom_revision: number;
  origin: "https://x.com"; account: string;
  state: "ready" | "login_required" | "challenge" | "unsupported";
};
export function readiness(value: unknown): value is Readiness {
  return p.object(value, ["binding_revision", "document", "dom_revision", "origin", "account", "state"])
    && p.id(value.binding_revision) && identity(value.document, 128) && p.counter(value.dom_revision)
    && value.origin === "https://x.com" && identity(value.account, 320)
    && ["ready", "login_required", "challenge", "unsupported"].includes(String(value.state));
}

File: crates/avesra-core/src/mailbox.rs
//! Bounded untrusted mailbox evidence. Callers must retain the actual browser
//! job; these data structures never grant authority or prove provider identity.
use avesra_contracts::ErrorCode;
use serde::Deserialize;
use std::{collections::HashSet, time::Instant};
use uuid::Uuid;

const MAX_BODY: usize = 16_384;
const MAX_BODIES: usize = 262_144;

#[derive(Clone, PartialEq, Eq, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Message {
    pub id: String,
    pub thread: String,
    pub timestamp_ms: u64,
    pub sender: String,
    pub subject: String,
    pub reference: String,
    pub body: String,
    pub body_complete: bool,
    pub thread_expanded: bool,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Batch {
    pub binding_revision: Uuid,
    pub account: String,
    pub scope: Scope,
    pub document: String,
    pub dom_revision: u64,
    pub ordinal: u16,
    pub cursor: Option<String>,
    pub next: Option<String>,
    pub end_of_inbox: bool,
    pub order: Order,
    pub messages: Vec<Message>,
}

#[derive(Clone, Copy, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Scope {
    Inbox,
}
#[derive(Clone, Copy, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Order {
    IndividualMessagesNewestFirst,
}

/// Intentionally neither Clone nor Serialize: message bodies stay within the
/// original native consumer. An incomplete accumulator cannot expose a report.
pub struct Mailbox {
    binding: Uuid,
    account: String,
    requested: usize,
    deadline: Instant,
    batches: u16,
    cursor: Option<String>,
    cursors: HashSet<String>,
    messages: Vec<Message>,
    bytes: usize,
    complete: bool,
    failed: bool,
}

fn text(value: &str, maximum: usize, multiline: bool) -> bool {
    value.len() <= maximum
        && value
            .chars()
            .all(|c| !c.is_control() || (multiline && matches!(c, '\n' | '\t')))
}
fn identity(value: &str, maximum: usize) -> bool {
    !value.is_empty() && value.len() <= maximum && value.bytes().all(|b| (33..=126).contains(&b))
}
impl Message {
    fn validate(&self) -> Result<(), ErrorCode> {
        if !identity(&self.id, 128)
            || !identity(&self.thread, 128)
            || !(1..=8_640_000_000_000_000).contains(&self.timestamp_ms)
            || self.sender.trim().is_empty()
            || !text(&self.sender, 320, false)
            || !text(&self.subject, 512, false)
            || !identity(&self.reference, 2048)
            || !text(&self.body, MAX_BODY, true)
            || !self.body_complete
            || !self.thread_expanded
        {
            return Err(ErrorCode::Malformed);
        }
        Ok(())
    }
}

impl Mailbox {
    pub fn new(
        binding: Uuid,
        account: String,
        requested: u16,
        original_deadline: Instant,
    ) -> Result<Self, ErrorCode> {
        if binding.is_nil() || !identity(&account, 320) || !(1..=100).contains(&requested) {
            return Err(ErrorCode::Malformed);
        }
        if Instant::now() >= original_deadline {
            return Err(ErrorCode::Expired);
        }
        Ok(Self {
            binding,
            account,
            requested: requested.into(),
            deadline: original_deadline,
            batches: 0,
            cursor: None,
            cursors: HashSet::new(),
            messages: Vec::new(),
            bytes: 0,
            complete: false,
            failed: false,
        })
    }

    /// The caller obtains Batch only from a validated owned provider operation.
    /// Any failed append poisons this accumulator; retry cannot hide a gap.
    pub fn append(&mut self, batch: Batch) -> Result<(), ErrorCode> {
        let result = self.append_inner(batch);
        if result.is_err() {
            self.failed = true;
            self.messages.clear();
        }
        result
    }
    fn append_inner(&mut self, batch: Batch) -> Result<(), ErrorCode> {
        if self.failed || self.complete || Instant::now() >= self.deadline {
            return Err(ErrorCode::Stale);
        }
        if self.batches >= 32
            || batch.ordinal != self.batches
            || batch.binding_revision != self.binding
            || batch.account != self.account
            || !identity(&batch.document, 128)
            || batch.dom_revision == 0
            || batch.dom_revision > 9_007_199_254_740_991
            || batch.cursor != self.cursor
            || batch.messages.len() > self.requested
            || batch.next.as_ref().is_some_and(|v| !identity(v, 256))
            || batch.end_of_inbox == batch.next.is_some()
            || (!batch.end_of_inbox && batch.messages.is_empty())
        {
            return Err(ErrorCode::Malformed);
        }
        if let Some(next) = &batch.next
            && !self.cursors.insert(next.clone())
        {
            return Err(ErrorCode::Stale);
        }
        let mut local_ids = HashSet::new();
        let mut previous = u64::MAX;
        for message in batch.messages {
            message.validate()?;
            if !local_ids.insert(message.id.clone()) || message.timestamp_ms > previous {
                return Err(ErrorCode::Malformed);
            }
            previous = message.timestamp_ms;
            if let Some(existing) = self.messages.iter().find(|v| v.id == message.id) {
                if existing != &message {
                    return Err(ErrorCode::Stale);
                }
                continue;
            }
            if self.messages.len() >= self.requested
                || self
                    .messages
                    .last()
                    .is_some_and(|v| v.timestamp_ms < message.timestamp_ms)
            {
                return Err(ErrorCode::Malformed);
            }
            self.bytes = self
                .bytes
                .checked_add(message.body.len())
                .ok_or(ErrorCode::TooLarge)?;
            if self.bytes > MAX_BODIES {
                return Err(ErrorCode::TooLarge);
            }
            self.messages.push(message);
        }
        if Instant::now() >= self.deadline {
            return Err(ErrorCode::Expired);
        }
        self.batches += 1;
        self.cursor = batch.next;
        self.complete = self.messages.len() == self.requested || batch.end_of_inbox;
        Ok(())
    }

    /// Borrowed complete messages are untrusted source material, never policy.
    /// The original browser owner must recheck authority before using a report.
    pub fn messages(&self) -> Result<&[Message], ErrorCode> {
        if self.failed || !self.complete || Instant::now() >= self.deadline {
            return Err(ErrorCode::Stale);
        }
        Ok(&self.messages)
    }
    pub fn actual_count(&self) -> Result<usize, ErrorCode> {
        self.messages().map(<[Message]>::len)
    }
}

/// Literal candidates only. A caller must preserve each message source and
/// ambiguity; these labels never attest physical delivery or package identity.
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum DeliveryClaim {
    Delivered,
    OutForDelivery,
    Shipped,
    Ambiguous,
}
pub struct Candidate<'a> {
    pub message: &'a Message,
    pub claim: DeliveryClaim,
    pub phrase: &'static str,
}
pub fn delivery_candidates(message: &Message) -> Result<Vec<Candidate<'_>>, ErrorCode> {
    message.validate()?;
    let combined = format!("{}\n{}", message.subject, message.body).to_ascii_lowercase();
    let mut candidates = Vec::new();
    for (phrase, claim) in [
        ("out for delivery", DeliveryClaim::OutForDelivery),
        ("has been delivered", DeliveryClaim::Delivered),
        ("was delivered", DeliveryClaim::Delivered),
        ("has shipped", DeliveryClaim::Shipped),
        ("has been shipped", DeliveryClaim::Shipped),
    ] {
        for (offset, _) in combined.match_indices(phrase) {
            let start = combined[..offset]
                .rfind(['.', '!', '?', '\n'])
                .map_or(0, |v| v + 1);
            let end = combined[offset..]
                .find(['.', '!', '?', '\n'])
                .map_or(combined.len(), |v| offset + v);
            let sentence = &combined[start..end];
            let uncertain = sentence.split(|c: char| !c.is_ascii_alphabetic()).any(|v| {
                matches!(
                    v,
                    "not"
                        | "never"
                        | "if"
                        | "estimated"
                        | "expected"
                        | "may"
                        | "might"
                        | "when"
                        | "unless"
                        | "isn"
                        | "hasn"
                        | "wasn"
                )
            });
            candidates.push(Candidate {
                message,
                claim: if uncertain {
                    DeliveryClaim::Ambiguous
                } else {
                    claim
                },
                phrase,
            });
            // A body is bounded, but do not expose an unbounded repetitive report.
            if candidates.len() == 8 {
                return Ok(candidates);
            }
        }
    }
    Ok(candidates)
}
